crates/bevy_ui/src/layout/mod.rs

```diff
@@ -241,93 +248,134 @@ pub enum LayoutError {
+            .push(entity);
+    }
+    let get_viewport_properties = |camera_id: Option<Entity>| {
+        if let Some(camera_id) = camera_id {
+            let camera = cameras.get(camera_id).unwrap();
+            let size = camera.physical_viewport_size().unwrap_or(UVec2::ZERO);
+            let scale_factor = camera.target_scaling_factor().unwrap_or(1.0) * ui_scale.0;
+            (size, scale_factor)
         } else {
-            return;
-        };
-
-    let resized = resize_events
-        .read()
-        .any(|resized_window| resized_window.window == primary_window_entity);
-
-    let scale_factor = logical_to_physical_factor * ui_scale.0;
+            let size = primary_window
+                .get_single()
+                .map(|(_, window)| [window.physical_width(), window.physical_height()].into())
+                .unwrap_or(UVec2::ZERO);
+            let scale_factor = primary_window
+                .get_single()
+                .map(|(_, window)| window.scale_factor())
+                .unwrap_or(1.0)
+                * ui_scale.0;
+            (size, scale_factor)
+        }
+    };
+
+    // Resize nodes
+    let resize_events = resize_events.read().collect::<Vec<_>>();
+    for (camera_id, root_nodes) in &camera_to_root_nodes_map {
+        let (size, scale_factor) = get_viewport_properties(*camera_id);
+        let resized = resize_events.iter().any(|resized_window| {
+            let primary_window = primary_window.get_single().map(|(e, _)| e).ok();
+            match camera_id {
+                Some(camera_id) => {
+                    if let Some(NormalizedRenderTarget::Window(window_ref)) = cameras
+                        .get(*camera_id)
+                        .ok()
+                        .and_then(|camera| camera.target.normalize(primary_window))
+                    {
+                        resized_window.window == window_ref.entity()
+                    } else {
+                        false
+                    }
+                }
+                None => {
+                    // If no camera specified, default to primary window
+                    match primary_window {
+                        Some(primary_window) => resized_window.window == primary_window,
+                        None => false,
+                    }
+                }
+            }
+        });
 
-    let layout_context = LayoutContext::new(scale_factor, physical_size);
+        let layout_context =
+            LayoutContext::new(scale_factor, [size.x as f32, size.y as f32].into());
 
-    if !scale_factor_events.is_empty() || ui_scale.is_changed() || resized {
-        scale_factor_events.clear();
         // update all nodes
-        for (entity, style) in style_query.iter() {
-            ui_surface.upsert_node(entity, &style, &layout_context);
-        }
-    } else {
-        for (entity, style) in style_query.iter() {
-            if style.is_changed() {
+        for (entity, style, ui_camera) in style_query.iter() {
```

Comment:
This iterates all UI nodes, for each UI camera. Would it make sense to only iterate once instead?

Reply:
I made a big change in recent commit https://github.com/bevyengine/bevy/pull/10559/commits/7676e63befd8df9b79dacefdd3b92fd2a42d15ab where I pre-calculate all relevant into for cameras first, and then either fetch it for each node, or iterate over. I wonder what you will think of it!